dropwizard-client 1.0.0rc1 can't handle multiple Set-Cookie headers

I'm having trouble with the implementation of the DropwizardApacheConnector. Specifically, its use of Collections.singletonList on line 98 of this file 

  https://github.com/dropwizard/dropwizard/blob/master/dropwizard-client/src/main/java/io/dropwizard/client/DropwizardApacheConnector.java

When receiveing a response with multiple cookies, they will all have a Set-Cookie header (with different data). Only the first such header will succeed. Subsequent headers can't be added since the headerValues, as far as I can tell, then always will be a Collections.singletonList, thus causing line 100 to fail since it can't add to such a "singleton" list. This results in headerValues.add causing an UnsupportedOperationException.

In order to move forward, I hacked around the problem for the time being as shown below. Please let me know if I'm totally missing the point here. If not, hopefully this helps helps someone else and shows a possible solution.

```
	for (Header header : apacheResponse.getAllHeaders()) {
		final List<String> headerValues = jerseyResponse.getHeaders().get(header.getName());
		if (headerValues == null) {
			List<String> list = new ArrayList<>(1);
			list.add(header.getValue());
			jerseyResponse.getHeaders().put(header.getName(), list);
		} else {
			headerValues.add(header.getValue());
		}
	}
```
-JM

the above test will currently fail with `java.lang.UnsupportedOperationException` - as there will be an attempt to add item to a `singletonList` (immutable).

https://github.com/dropwizard/dropwizard/blob/master/dropwizard-client/src/main/java/io/dropwizard/client/DropwizardApacheConnector.java#L99

Full stacktrace:

```
javax.ws.rs.ProcessingException: java.lang.UnsupportedOperationException
	at io.dropwizard.client.DropwizardApacheConnector.apply(DropwizardApacheConnector.java:109)
	at io.dropwizard.client.DropwizardApacheConnectorTest.applyShouldHandleMultipleHeadersWithTheSameKey(DropwizardApacheConnectorTest.java:195)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.intellij.junit4.JUnit4TestRunnerUtil$IgnoreIgnoredTestJUnit4ClassRunner.runChild(JUnit4TestRunnerUtil.java:341)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractList.add(AbstractList.java:148)
	at java.util.AbstractList.add(AbstractList.java:108)
	at io.dropwizard.client.DropwizardApacheConnector.apply(DropwizardApacheConnector.java:99)
	... 32 more
```

--- a/dropwizard-client/src/main/java/io/dropwizard/client/DropwizardApacheConnector.java
+++ b/dropwizard-client/src/main/java/io/dropwizard/client/DropwizardApacheConnector.java
@@ -1,5 +1,6 @@
 package io.dropwizard.client;
 
+import com.google.common.collect.Lists;
 import com.google.common.util.concurrent.MoreExecutors;
 import org.apache.http.Header;
 import org.apache.http.HttpEntity;
@@ -26,7 +27,6 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
-import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.Future;
@@ -94,7 +94,7 @@ public class DropwizardApacheConnector implements Connector {
             for (Header header : apacheResponse.getAllHeaders()) {
                 final List<String> headerValues = jerseyResponse.getHeaders().get(header.getName());
                 if (headerValues == null) {
-                    jerseyResponse.getHeaders().put(header.getName(), Collections.singletonList(header.getValue()));
+                    jerseyResponse.getHeaders().put(header.getName(), Lists.newArrayList(header.getValue()));
                 } else {
                     headerValues.add(header.getValue());
                 }

--- a/dropwizard-client/src/test/java/io/dropwizard/client/DropwizardApacheConnectorTest.java
+++ b/dropwizard-client/src/test/java/io/dropwizard/client/DropwizardApacheConnectorTest.java
@@ -10,12 +10,20 @@ import io.dropwizard.setup.Environment;
 import io.dropwizard.testing.ResourceHelpers;
 import io.dropwizard.testing.junit.DropwizardAppRule;
 import io.dropwizard.util.Duration;
+import org.apache.http.Header;
 import org.apache.http.HttpStatus;
+import org.apache.http.ProtocolVersion;
+import org.apache.http.client.methods.CloseableHttpResponse;
 import org.apache.http.conn.ConnectTimeoutException;
 import org.apache.http.conn.HttpHostConnectException;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.message.BasicStatusLine;
 import org.assertj.core.api.AbstractLongAssert;
 import org.eclipse.jetty.util.component.LifeCycle;
 import org.glassfish.jersey.client.ClientProperties;
+import org.glassfish.jersey.client.ClientRequest;
+import org.glassfish.jersey.client.ClientResponse;
 import org.glassfish.jersey.client.JerseyClient;
 import org.junit.After;
 import org.junit.Before;
@@ -24,11 +32,13 @@ import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
+import org.mockito.Matchers;
 
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 import javax.ws.rs.ProcessingException;
 import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.MultivaluedHashMap;
 import javax.ws.rs.core.Response;
 import java.net.NoRouteToHostException;
 import java.net.SocketTimeoutException;
@@ -37,6 +47,8 @@ import java.util.concurrent.TimeUnit;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.hamcrest.CoreMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 @Ignore //These tests are consistently failing on travis CI because of network timeouts
 public class DropwizardApacheConnectorTest {
@@ -163,6 +175,32 @@ public class DropwizardApacheConnectorTest {
                     .get(String.class);
             assertThat(response).isEqualTo("success");
         }
+    }
+
+    @Test
+    public void multiple_headers_with_the_same_name_are_processed_successfully() throws Exception {
+
+        final CloseableHttpClient client = mock(CloseableHttpClient.class);
+        final DropwizardApacheConnector dropwizardApacheConnector = new DropwizardApacheConnector(client, null, false);
+        final Header[] apacheHeaders = {
+            new BasicHeader("Set-Cookie", "test1"),
+            new BasicHeader("Set-Cookie", "test2")
+        };
+
+        final CloseableHttpResponse apacheResponse = mock(CloseableHttpResponse.class);
+        when(apacheResponse.getStatusLine()).thenReturn(new BasicStatusLine(new ProtocolVersion("HTTP", 1, 1), 200, "OK"));
+        when(apacheResponse.getAllHeaders()).thenReturn(apacheHeaders);
+        when(client.execute(Matchers.any())).thenReturn(apacheResponse);
+
+        final ClientRequest jerseyRequest = mock(ClientRequest.class);
+        when(jerseyRequest.getUri()).thenReturn(URI.create("http://localhost"));
+        when(jerseyRequest.getMethod()).thenReturn("GET");
+        when(jerseyRequest.getHeaders()).thenReturn(new MultivaluedHashMap<>());
+
+        final ClientResponse jerseyResponse = dropwizardApacheConnector.apply(jerseyRequest);
+
+        assertThat(jerseyResponse.getStatus()).isEqualTo(apacheResponse.getStatusLine().getStatusCode());
+
     }
 
     @Path("/")

